hooks/mari/publish_mipmaps.py

```diff
@@ -0,0 +1,221 @@
+            source_path = item.parent.properties.publish_path
+            seq_path = publisher.util.get_frame_sequence_path(source_path)
+            if seq_path:
+                src_files = publisher.util.get_sequence_path_files(source_path)
+            else:
+                src_files = [source_path]
+
+            # write mipmaps for each file in the sequence
+            self.create_mipmaps_for_seq(src_files, target_path)
+
+        except Exception as e:
+            raise TankError("Failed to publish file for item '%s': %s" % (item.name, str(e)))
+
+        self.logger.debug(
+            "Published file for item '%s' to '%s'." % (item.name, target_path)
+        )
+
+        return [target_path]
+
+    def create_mipmaps_for_seq(self, source_paths, target_seq_path):
+        """
+        Given a list of files, convert each to a mipmap file and save to target path.
+
+        :param source_paths:        list of source image files
+        :param target_seq_path:     path (with an seq field if source is an image sequence)
+                                    where the mipmaps should be written
+        """
+        publisher = self.parent
+
+        for source_path in source_paths:
+            frame = publisher.util.get_frame_number(source_path)
+            if frame:
+                target_path = publisher.util.get_path_for_frame(target_seq_path, frame)
+                if not target_path:
+                    # We do not want the conversion of multiple images to overwrite
+                    # a single target path. Something is wrong with the configuration.
+                    raise TankError("Source path: {} contains a frame number, "
+                                    "but target path: {} does not.".format(source_path, target_path))
+            else:
+                target_path = target_seq_path
+
+            if not self._create_mipmap(source_path, target_path):
+                self.logger.warning("Mipmap not created for: {}".format(target_path))
+
+    def _create_mipmap(self, source_path, target_path):
+        """
+        Use OIIO to convert a given image into a mipmapped image.
+
+        :param source_path: path to source image file
+        :param target_path: path to write the mipmapped file to
+
+        :return: bool (success of mipmap creation)
+        """
+        # cast here as OIIO has an issue with unicode strings (C++ matches types strictly)
+        source_path = str(source_path)
+        target_path = str(target_path)
+
+        if not self._valid_for_mipmap_multiimage(target_path):
+            self.logger.warning("Mipmap not created for: {}".format(target_path))
+            return False
+
+        _img_input = oiio.ImageBuf(source_path)
+        _target_spec = oiio.ImageSpec(_img_input.spec())
+        _target_spec.attribute("maketx:filtername", "lanczos3")
+
+        return oiio.ImageBufAlgo.make_texture(oiio.MakeTxTexture, _img_input, target_path, _target_spec)
```

Comment:
This procedure needs to be derived from the file extension. I suspect `oiio.MakeTxTexture` wouldn't work for, say, a `.rat` file. We should raise an exception of some sort, `TypeError` at the very least, or perhaps a custom `UnsupportedFileType` exception, for unsupported file types.

Reply:
Additionally, we should register supported extensions, and then catch this in the validation step, rather than waiting until publish to fail.

Reply:
Or perhaps even the acceptance step. That seems most appropriate. Accept the plugin, so it still shows in the UI, but disable it so it doesn't run.

Reply:
Hm. Agreed about validation, failing at publish time is not at all good...let me look at that.

Do we even want to think about having file type specific plugins? I mean, in case we want to use a renderer specific tool to create the mipmap, instead of oiio...or we can look at it in the future.

Reply:
Re. accept and disable: 
- if I do this, I think artists are likely to ignore it (see screenshot)
![accept_and_disable](https://user-images.githubusercontent.com/35912094/51379803-47e6c200-1b36-11e9-8db9-a9d11a4739b1.png)
Since there is no actual warning sign (things appear in more info, but no one checks that), I think people will just continue publishing and later complain that the files were not generated
- I would have to get publish_path (to pass to `_valid_for_mipmap_multiimage()`. This property is being set (and validated) in the validate method currently. I wouldn't like to change it only for this plugin.

Reply:
I have updated the code with the check happening in validate (not in accept). I think that is more likely to get the issue reported...let me know what you think.

Reply:
Its not really a validation error though, its a configuration issue. i.e. its not something that an artist should fix or even care about. If you are concerned about cruft in the UI, then I would just fail acceptance entirely.

As a side note, we should probably embed the extension in the mipmap task name so we can tell them apart if we end up generating more than one.

Reply:
I'm not concerned about cruft in the UI. Though it is not an issue that should be fixed by the artist, I still think it is an issue that should be noticed by the artist and reported to a TD.

Extension is at the end of the mipmap task name...unseen in the screenshot because of the long names. The geo names are ugly long, but I don't think that can be helped.

I created new items because without a separate item, it wouldn't be possible to maintain a separate publish_path (and template) I could add these as separate properties to `mari.channel` and `mari.texture` items, and rewrite `publish()` (bit of code repetition, but I guess that's okay).

Reply:
publish_path is defined on the Plugin (i.e. Task) not the Item, so you should be fine there. I would still maintain a different Task instance for each mipmap extension, but all just derived from one shared Plugin def.

Reply:
There is no code change required for this, just settings. Just update the tk-multi-publish config block to include multiple instances of the "mipmap" plugin, with the distinction of each instance specifying a particular "mode" it will operate in (i.e. "tx mode", "rat mode", etc).

Reply:
Both are using the `@settings.tk-multi-publish2.mari.publish_mipmaps` definition, the only difference is the `Publish Extension` setting.